src/customRules/structuredImportsRule.ts

```diff
@@ -0,0 +1,94 @@
+import * as ts from 'typescript';
+import * as Lint from 'tslint';
+
+export class Rule extends Lint.Rules.AbstractRule {
+  /* tslint:disable:object-literal-sort-keys */
+  public static metadata: Lint.IRuleMetadata = {
+    'ruleName': 'structured-imports',
+    'description': 'Enforce structure to your imports. Import structure should be listed in the folloing order: modules, absolute imports,  relative parent/ancestor directories, relative sibling directors.',
+    'hasFix': false,
+    'optionsDescription': 'Not configurable.',
+    'options': null,
+    'optionExamples': null,
+    'type': 'style',
+    'typescriptOnly': false,
+  };
+  /* tslint:enable:object-literal-sort-keys */
+
+  public static STRUCTURED_IMPORTS_ABS_FIRST_ERROR = 'Imports should be listed in the following order: module imports, absolute imports, relative imports.';
+
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    const structuredImportsWalker = new StructuredImportsWalker(sourceFile, this.getOptions());
+    return this.applyWithWalker(structuredImportsWalker);
+  }
+}
+
+enum ImportType {
+  Module = 0, // 'module'
+  Absolute = 1, // '/some/folder/file'
+  RelativeAncestor = 2, // '../parentFolder'
+  RelativeSibling = 3, // './siblingFolder'
+};
+
+const importStuctureOrder = [ImportType.Module, ImportType.Absolute, ImportType.RelativeAncestor, ImportType.RelativeSibling];
+
+class StructuredImportsWalker extends Lint.RuleWalker {
+  private previousImport: ImportType;
+  private currentImport: ImportType;
+
+  constructor(sourceFile: ts.SourceFile, options: Lint.IOptions) {
+    super(sourceFile, options);
+  }
+
+  // e.g. 'import Foo from './foo';'
+  public visitImportDeclaration(node: ts.ImportDeclaration) {
```

Comment:
So normally this would be cool, but I'm really concerned about performance.  `yarn check` is taking over 2 minutes to run (>40% of CI time), and that'll only get worse as we add more code + rules.  AFAIK, there's no way to parallelize tslint.

Taking a look at the [tslint performance guide](https://palantir.github.io/tslint/develop/custom-rules/performance.html), it recommends:
- _Don’t walk the whole AST if possible_
- _Implement your own walking algorithm_

> If you are interested in import statements for example, you only need to search in sourceFile.statements and nested NamespaceDeclaration

... and since nested declarations can't do module imports, that limits this rule to `sourceFile.statements`.

Here's a rough version of the above, with annotations on how to improve it.  It passes for the existing tests, at least.

```ts
// This should be adapted to `extend Lint.AbstractWalker<MyOptionsType>`, and implement the `walk` rule.
class StructuredImportsWalker extends Lint.RuleWalker {
  private previousImport: ImportType;
  private currentImport: ImportType;

  constructor(sourceFile: ts.SourceFile, options: Lint.IOptions) {
    super(sourceFile, options);
  }

  public visitSourceFile(node: ts.SourceFile) {
    const importNodes = node.statements
      .filter((child) => child.kind === ts.SyntaxKind.ImportDeclaration)
      .map((child) => child as ts.ImportDeclaration);

    if (importNodes.length === 0) { return; }

    // This is off the top of my head; probably a more readable way to do this.
    let lastGoodType = getImportType(importNodes.shift());
    while (importNodes.length) {
      const currentImport = importNodes.shift();
      const currentType = getImportType(currentImport);
      // Here we're just relying on enum comparisons.  This gets rid of a couple of helper functions.
      if (lastGoodType > currentType) {
        this.addFailure(
          this.createFailure(
            currentImport.moduleSpecifier.getStart(), 
            currentImport.moduleSpecifier.getFullWidth(),
            Rule.STRUCTURED_IMPORTS_ABS_FIRST_ERROR,
          )
        );
      } else {
        lastGoodType = currentType;
      }
    }
  }
}
```